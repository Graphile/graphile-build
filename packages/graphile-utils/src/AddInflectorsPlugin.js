export default function AddInflectorsPlugin(additionalInflectors) {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      return build.extend(inflection, additionalInflectors);
    });
  };
}
