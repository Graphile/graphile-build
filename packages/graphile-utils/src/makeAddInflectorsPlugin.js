export default function makeAddInflectorsPlugin(additionalInflectors) {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      return build.extend(inflection, additionalInflectors);
    });
  };
}
