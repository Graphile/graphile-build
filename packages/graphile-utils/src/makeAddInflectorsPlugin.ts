import { Plugin } from "graphile-build";

export default function makeAddInflectorsPlugin(
  additionalInflectors: {
    [str: string]: ((...args: Array<any>) => any);
  },
  replace = false
): Plugin {
  return builder => {
    builder.hook("inflection", (inflection, build) => {
      if (replace) {
        // Overwrite directly so that we don't lose the 'extend' hints
        Object.assign(inflection, additionalInflectors);
        return inflection;
      } else {
        return build.extend(
          inflection,
          additionalInflectors,
          `Adding inflectors ('${Object.keys(additionalInflectors).join(
            "', '"
          )}') via makeAddInflectorsPlugin. You can pass \`true\` as the second argument to makeAddInflectorsPlugin to allow overwriting existing inflectors.`
        );
      }
    });
  };
}
