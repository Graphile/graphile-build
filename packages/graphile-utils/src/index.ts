import { embed, gql } from "./gql";
import makeAddInflectorsPlugin from "./makeAddInflectorsPlugin";
import makeExtendSchemaPlugin from "./makeExtendSchemaPlugin";
import makePluginByCombiningPlugins from "./makePluginByCombiningPlugins";
import makeWrapResolversPlugin from "./makeWrapResolversPlugin";

export {
  embed,
  gql,
  makeAddInflectorsPlugin,
  makeExtendSchemaPlugin,
  makePluginByCombiningPlugins,
  makeWrapResolversPlugin,
};
