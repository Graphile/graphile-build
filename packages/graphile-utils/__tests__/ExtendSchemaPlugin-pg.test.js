import pg from "pg";
import { graphql, printSchema } from "graphql";
import { createPostGraphileSchema } from "postgraphile-core";
import { ExtendSchemaPlugin, gql } from "../src";

const clean = data => {
  if (Array.isArray(data)) {
    return data.map(clean);
  } else if (data && typeof data === "object") {
    return Object.keys(data).reduce((memo, key) => {
      const value = data[key];
      if (key === "id" && typeof value === "number") {
        memo[key] = "[id]";
      } else if (key === "nodeId" && typeof value === "string") {
        memo[key] = "[nodeId]";
      } else {
        memo[key] = clean(value);
      }
      return memo;
    }, {});
  } else {
    return data;
  }
};

function mockSendEmail() {
  return new Promise(resolve => setTimeout(resolve, 1));
}
let pgPool;

beforeAll(() => {
  pgPool = new pg.Pool({
    connectionString: process.env.TEST_DATABASE_URL,
  });
});

afterAll(() => {
  if (pgPool) {
    pgPool.end();
    pgPool = null;
  }
});

it("allows adding a custom single field to PG schema", async () => {
  const schema = await createPostGraphileSchema(pgPool, ["a"], {
    disableDefaultMutations: true,
    appendPlugins: [
      ExtendSchemaPlugin(build => {
        const { pgSql: sql } = build;
        return {
          typeDefs: gql`
            extend type Query {
              randomUser: User
            }
          `,
          resolvers: {
            Query: {
              async randomUser(_query, args, context, resolveInfo, { selectGraphQLResultFromTable }) {
                const rows = await selectGraphQLResultFromTable(
                  sql.fragment`a.users`,
                  (tableAlias, sqlBuilder) => {
                    sqlBuilder.orderBy(sql.fragment`random()`);
                    sqlBuilder.limit(1);
                  }
                );
                return rows[0];
              },
            },
          },
        };
      }),
    ],
  });
  const printedSchema = printSchema(schema);
  expect(printedSchema).toMatchSnapshot();
  const pgClient = await pgPool.connect();
  try {
    const { data, errors } = await graphql(
      schema,
      `
        query {
          randomUser {
            nodeId
            id
            name
            email
            bio
          }
        }
      `,
      null,
      { pgClient },
      {}
    );
    expect(errors).toBeFalsy();
    expect(data).toBeTruthy();
    expect(data.randomUser).toBeTruthy();
    expect(data.randomUser.id).toBeTruthy();
    expect(data.randomUser.nodeId).toBeTruthy();
    expect(data.randomUser.name).toBeTruthy();
    expect(data.randomUser.email).toBeTruthy();
  } finally {
    pgClient.release();
  }
});

it("allows adding a custom field returning a list to PG schema", async () => {
  const schema = await createPostGraphileSchema(pgPool, ["a"], {
    disableDefaultMutations: true,
    appendPlugins: [
      ExtendSchemaPlugin(build => {
        const { pgSql: sql } = build;
        return {
          typeDefs: gql`
            extend type Query {
              randomUsers: [User!]
            }
          `,
          resolvers: {
            Query: {
              async randomUsers(
                _query,
                args,
                context,
                resolveInfo,
                { selectGraphQLResultFromTable }
              ) {
                const rows = await selectGraphQLResultFromTable(
                  sql.fragment`a.users`,
                  (tableAlias, sqlBuilder) => {
                    sqlBuilder.orderBy(sql.fragment`random()`);
                    sqlBuilder.limit(3);
                  }
                );
                return rows;
              },
            },
          },
        };
      }),
    ],
  });
  const printedSchema = printSchema(schema);
  expect(printedSchema).toMatchSnapshot();
  const pgClient = await pgPool.connect();
  try {
    const { data, errors } = await graphql(
      schema,
      `
        query {
          randomUsers {
            nodeId
            id
            name
            email
            bio
          }
        }
      `,
      null,
      { pgClient },
      {}
    );
    expect(errors).toBeFalsy();
    expect(data).toBeTruthy();
    expect(data.randomUsers).toBeTruthy();
    expect(data.randomUsers.length).toEqual(3);
    expect(data.randomUsers[2].id).toBeTruthy();
    expect(data.randomUsers[2].nodeId).toBeTruthy();
    expect(data.randomUsers[2].name).toBeTruthy();
    expect(data.randomUsers[2].email).toBeTruthy();
  } finally {
    pgClient.release();
  }
});

it("allows adding a simple mutation field to PG schema", async () => {
  const schema = await createPostGraphileSchema(pgPool, ["a"], {
    disableDefaultMutations: true,
    appendPlugins: [
      ExtendSchemaPlugin(build => {
        const { pgSql: sql } = build;
        return {
          typeDefs: gql`
            input RegisterUserInput {
              name: String!
              email: String!
              bio: String
            }

            type RegisterUserPayload {
              user: User @recurseDataGenerators
            }

            extend type Mutation {
              registerUser(input: RegisterUserInput!): RegisterUserPayload
            }
          `,
          resolvers: {
            Mutation: {
              async registerUser(
                _query,
                args,
                context,
                resolveInfo,
                { selectGraphQLResultFromTable }
              ) {
                const { pgClient } = context;
                await pgClient.query("begin");
                try {
                  const { rows: [user] } = await pgClient.query(
                    `insert into a.users(name, email, bio) values ($1, $2, $3) returning *`,
                    [args.input.name, args.input.email, args.input.bio]
                  );
                  const [row] = await selectGraphQLResultFromTable(
                    sql.fragment`a.users`,
                    (tableAlias, sqlBuilder) => {
                      sqlBuilder.where(
                        sql.fragment`${tableAlias}.id = ${sql.value(user.id)}`
                      );
                    }
                  );
                  await mockSendEmail(
                    args.input.email,
                    "Welcome to my site",
                    `You're user ${user.id} - thanks for being awesome`
                  );

                  await pgClient.query("commit");
                  return {
                    user: row,
                  };
                } catch (e) {
                  await pgClient.query("rollback");
                  throw e;
                }
              },
            },
          },
        };
      }),
    ],
  });
  const printedSchema = printSchema(schema);
  expect(printedSchema).toMatchSnapshot();
  const pgClient = await pgPool.connect();
  try {
    const { data, errors } = await graphql(
      schema,
      `
        mutation {
          user1: registerUser(
            input: { name: "Test User 1", email: "testuser1@example.com" }
          ) {
            user {
              nodeId
              id
              name
              email
              bio
            }
          }
          user2: registerUser(
            input: {
              name: "Test User 2"
              email: "testuser2@example.com"
              bio: "I have a bio!"
            }
          ) {
            user {
              nodeId
              id
              name
              email
              bio
            }
          }
        }
      `,
      null,
      { pgClient },
      {}
    );
    expect(errors).toBeFalsy();
    expect(data.user1).toBeTruthy();
    expect(data.user1.user.nodeId).toBeTruthy();
    expect(data.user1.user.id).toBeTruthy();
    expect(data.user2.user.nodeId).toBeTruthy();
    expect(data.user2.user.id).toBeTruthy();
    expect(data.user1.user.id).not.toEqual(data.user2.user.id);
    expect(clean(data)).toMatchSnapshot();
  } finally {
    pgClient.release();
  }
});
