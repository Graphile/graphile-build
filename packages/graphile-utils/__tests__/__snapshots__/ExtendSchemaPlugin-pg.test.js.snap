// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows adding a custom field returning a list to PG schema 1`] = `
"type Complex {
  numberInt: Int
  stringText: String
}

\\"\\"\\"An input for mutations affecting \`Complex\`\\"\\"\\"
input ComplexInput {
  numberInt: Int
  stringText: String
}

\\"\\"\\"A location in a connection that can be used for resuming pagination.\\"\\"\\"
scalar Cursor

\\"\\"\\"
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
\\"\\"\\"
scalar Datetime

\\"\\"\\"An object with a globally unique \`ID\`.\\"\\"\\"
interface Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: Cursor

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: Cursor
}

\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query implements Node {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  \\"\\"\\"
  nodeId: ID!

  \\"\\"\\"Fetches an object given its globally unique \`ID\`.\\"\\"\\"
  node(
    \\"\\"\\"The globally unique \`ID\`.\\"\\"\\"
    nodeId: ID!
  ): Node

  \\"\\"\\"Reads and enables pagination through a set of \`User\`.\\"\\"\\"
  allUsers(
    \\"\\"\\"Only read the first \`n\` values of the set.\\"\\"\\"
    first: Int

    \\"\\"\\"Only read the last \`n\` values of the set.\\"\\"\\"
    last: Int

    \\"\\"\\"
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    \\"\\"\\"
    offset: Int

    \\"\\"\\"Read all values in the set before (above) this cursor.\\"\\"\\"
    before: Cursor

    \\"\\"\\"Read all values in the set after (below) this cursor.\\"\\"\\"
    after: Cursor

    \\"\\"\\"The method to use when ordering \`User\`.\\"\\"\\"
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    \\"\\"\\"
    A condition to be used in determining which values should be returned by the collection.
    \\"\\"\\"
    condition: UserCondition
  ): UsersConnection
  userById(id: Int!): User

  \\"\\"\\"Reads a single \`User\` using its globally unique \`ID\`.\\"\\"\\"
  user(
    \\"\\"\\"The globally unique \`ID\` to be used in selecting a single \`User\`.\\"\\"\\"
    nodeId: ID!
  ): User
  randomUsers: [User!]
}

type User implements Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
}

\\"\\"\\"
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
\\"\\"\\"
input UserCondition {
  \\"\\"\\"Checks for equality with the object’s \`id\` field.\\"\\"\\"
  id: Int

  \\"\\"\\"Checks for equality with the object’s \`name\` field.\\"\\"\\"
  name: String

  \\"\\"\\"Checks for equality with the object’s \`email\` field.\\"\\"\\"
  email: String

  \\"\\"\\"Checks for equality with the object’s \`bio\` field.\\"\\"\\"
  bio: String

  \\"\\"\\"Checks for equality with the object’s \`renamedComplexColumn\` field.\\"\\"\\"
  renamedComplexColumn: [ComplexInput]

  \\"\\"\\"Checks for equality with the object’s \`createdAt\` field.\\"\\"\\"
  createdAt: Datetime
}

\\"\\"\\"A connection to a list of \`User\` values.\\"\\"\\"
type UsersConnection {
  \\"\\"\\"A list of \`User\` objects.\\"\\"\\"
  nodes: [User]!

  \\"\\"\\"
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  \\"\\"\\"
  edges: [UsersEdge!]!

  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"The count of *all* \`User\` you could get from the connection.\\"\\"\\"
  totalCount: Int
}

\\"\\"\\"A \`User\` edge in the connection.\\"\\"\\"
type UsersEdge {
  \\"\\"\\"A cursor for use in pagination.\\"\\"\\"
  cursor: Cursor

  \\"\\"\\"The \`User\` at the end of the edge.\\"\\"\\"
  node: User
}

\\"\\"\\"Methods to use when ordering \`User\`.\\"\\"\\"
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}
"
`;

exports[`allows adding a custom single field to PG schema 1`] = `
"type Complex {
  numberInt: Int
  stringText: String
}

\\"\\"\\"An input for mutations affecting \`Complex\`\\"\\"\\"
input ComplexInput {
  numberInt: Int
  stringText: String
}

\\"\\"\\"A location in a connection that can be used for resuming pagination.\\"\\"\\"
scalar Cursor

\\"\\"\\"
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
\\"\\"\\"
scalar Datetime

\\"\\"\\"An object with a globally unique \`ID\`.\\"\\"\\"
interface Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: Cursor

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: Cursor
}

\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query implements Node {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  \\"\\"\\"
  nodeId: ID!

  \\"\\"\\"Fetches an object given its globally unique \`ID\`.\\"\\"\\"
  node(
    \\"\\"\\"The globally unique \`ID\`.\\"\\"\\"
    nodeId: ID!
  ): Node

  \\"\\"\\"Reads and enables pagination through a set of \`User\`.\\"\\"\\"
  allUsers(
    \\"\\"\\"Only read the first \`n\` values of the set.\\"\\"\\"
    first: Int

    \\"\\"\\"Only read the last \`n\` values of the set.\\"\\"\\"
    last: Int

    \\"\\"\\"
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    \\"\\"\\"
    offset: Int

    \\"\\"\\"Read all values in the set before (above) this cursor.\\"\\"\\"
    before: Cursor

    \\"\\"\\"Read all values in the set after (below) this cursor.\\"\\"\\"
    after: Cursor

    \\"\\"\\"The method to use when ordering \`User\`.\\"\\"\\"
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    \\"\\"\\"
    A condition to be used in determining which values should be returned by the collection.
    \\"\\"\\"
    condition: UserCondition
  ): UsersConnection
  userById(id: Int!): User

  \\"\\"\\"Reads a single \`User\` using its globally unique \`ID\`.\\"\\"\\"
  user(
    \\"\\"\\"The globally unique \`ID\` to be used in selecting a single \`User\`.\\"\\"\\"
    nodeId: ID!
  ): User
  randomUser: User
}

type User implements Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
}

\\"\\"\\"
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
\\"\\"\\"
input UserCondition {
  \\"\\"\\"Checks for equality with the object’s \`id\` field.\\"\\"\\"
  id: Int

  \\"\\"\\"Checks for equality with the object’s \`name\` field.\\"\\"\\"
  name: String

  \\"\\"\\"Checks for equality with the object’s \`email\` field.\\"\\"\\"
  email: String

  \\"\\"\\"Checks for equality with the object’s \`bio\` field.\\"\\"\\"
  bio: String

  \\"\\"\\"Checks for equality with the object’s \`renamedComplexColumn\` field.\\"\\"\\"
  renamedComplexColumn: [ComplexInput]

  \\"\\"\\"Checks for equality with the object’s \`createdAt\` field.\\"\\"\\"
  createdAt: Datetime
}

\\"\\"\\"A connection to a list of \`User\` values.\\"\\"\\"
type UsersConnection {
  \\"\\"\\"A list of \`User\` objects.\\"\\"\\"
  nodes: [User]!

  \\"\\"\\"
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  \\"\\"\\"
  edges: [UsersEdge!]!

  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"The count of *all* \`User\` you could get from the connection.\\"\\"\\"
  totalCount: Int
}

\\"\\"\\"A \`User\` edge in the connection.\\"\\"\\"
type UsersEdge {
  \\"\\"\\"A cursor for use in pagination.\\"\\"\\"
  cursor: Cursor

  \\"\\"\\"The \`User\` at the end of the edge.\\"\\"\\"
  node: User
}

\\"\\"\\"Methods to use when ordering \`User\`.\\"\\"\\"
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}
"
`;

exports[`allows adding a field to an existing table, and requesting necessary data along with it 1`] = `
"type Complex {
  numberInt: Int
  stringText: String
}

\\"\\"\\"An input for mutations affecting \`Complex\`\\"\\"\\"
input ComplexInput {
  numberInt: Int
  stringText: String
}

\\"\\"\\"A location in a connection that can be used for resuming pagination.\\"\\"\\"
scalar Cursor

\\"\\"\\"
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
\\"\\"\\"
scalar Datetime

\\"\\"\\"An object with a globally unique \`ID\`.\\"\\"\\"
interface Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: Cursor

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: Cursor
}

\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query implements Node {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  \\"\\"\\"
  nodeId: ID!

  \\"\\"\\"Fetches an object given its globally unique \`ID\`.\\"\\"\\"
  node(
    \\"\\"\\"The globally unique \`ID\`.\\"\\"\\"
    nodeId: ID!
  ): Node

  \\"\\"\\"Reads and enables pagination through a set of \`User\`.\\"\\"\\"
  allUsers(
    \\"\\"\\"Only read the first \`n\` values of the set.\\"\\"\\"
    first: Int

    \\"\\"\\"Only read the last \`n\` values of the set.\\"\\"\\"
    last: Int

    \\"\\"\\"
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    \\"\\"\\"
    offset: Int

    \\"\\"\\"Read all values in the set before (above) this cursor.\\"\\"\\"
    before: Cursor

    \\"\\"\\"Read all values in the set after (below) this cursor.\\"\\"\\"
    after: Cursor

    \\"\\"\\"The method to use when ordering \`User\`.\\"\\"\\"
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    \\"\\"\\"
    A condition to be used in determining which values should be returned by the collection.
    \\"\\"\\"
    condition: UserCondition
  ): UsersConnection
  userById(id: Int!): User

  \\"\\"\\"Reads a single \`User\` using its globally unique \`ID\`.\\"\\"\\"
  user(
    \\"\\"\\"The globally unique \`ID\` to be used in selecting a single \`User\`.\\"\\"\\"
    nodeId: ID!
  ): User
}

type User implements Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
  customField: String
}

\\"\\"\\"
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
\\"\\"\\"
input UserCondition {
  \\"\\"\\"Checks for equality with the object’s \`id\` field.\\"\\"\\"
  id: Int

  \\"\\"\\"Checks for equality with the object’s \`name\` field.\\"\\"\\"
  name: String

  \\"\\"\\"Checks for equality with the object’s \`email\` field.\\"\\"\\"
  email: String

  \\"\\"\\"Checks for equality with the object’s \`bio\` field.\\"\\"\\"
  bio: String

  \\"\\"\\"Checks for equality with the object’s \`renamedComplexColumn\` field.\\"\\"\\"
  renamedComplexColumn: [ComplexInput]

  \\"\\"\\"Checks for equality with the object’s \`createdAt\` field.\\"\\"\\"
  createdAt: Datetime
}

\\"\\"\\"A connection to a list of \`User\` values.\\"\\"\\"
type UsersConnection {
  \\"\\"\\"A list of \`User\` objects.\\"\\"\\"
  nodes: [User]!

  \\"\\"\\"
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  \\"\\"\\"
  edges: [UsersEdge!]!

  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"The count of *all* \`User\` you could get from the connection.\\"\\"\\"
  totalCount: Int
}

\\"\\"\\"A \`User\` edge in the connection.\\"\\"\\"
type UsersEdge {
  \\"\\"\\"A cursor for use in pagination.\\"\\"\\"
  cursor: Cursor

  \\"\\"\\"The \`User\` at the end of the edge.\\"\\"\\"
  node: User
}

\\"\\"\\"Methods to use when ordering \`User\`.\\"\\"\\"
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}
"
`;

exports[`allows adding a simple mutation field to PG schema 1`] = `
"type Complex {
  numberInt: Int
  stringText: String
}

\\"\\"\\"An input for mutations affecting \`Complex\`\\"\\"\\"
input ComplexInput {
  numberInt: Int
  stringText: String
}

\\"\\"\\"A location in a connection that can be used for resuming pagination.\\"\\"\\"
scalar Cursor

\\"\\"\\"
A point in time as described by the [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) standard. May or may not include a timezone.
\\"\\"\\"
scalar Datetime

\\"\\"\\"
The root mutation type which contains root level fields which mutate data.
\\"\\"\\"
type Mutation {
  registerUser(
    \\"\\"\\"
    The exclusive input argument for this mutation. An object type, make sure to see documentation for this object’s fields.
    \\"\\"\\"
    input: RegisterUserInput!
  ): RegisterUserPayload
}

\\"\\"\\"An object with a globally unique \`ID\`.\\"\\"\\"
interface Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
}

\\"\\"\\"Information about pagination in a connection.\\"\\"\\"
type PageInfo {
  \\"\\"\\"When paginating forwards, are there more items?\\"\\"\\"
  hasNextPage: Boolean!

  \\"\\"\\"When paginating backwards, are there more items?\\"\\"\\"
  hasPreviousPage: Boolean!

  \\"\\"\\"When paginating backwards, the cursor to continue.\\"\\"\\"
  startCursor: Cursor

  \\"\\"\\"When paginating forwards, the cursor to continue.\\"\\"\\"
  endCursor: Cursor
}

\\"\\"\\"The root query type which gives access points into the data universe.\\"\\"\\"
type Query implements Node {
  \\"\\"\\"
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  \\"\\"\\"
  query: Query!

  \\"\\"\\"
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  \\"\\"\\"
  nodeId: ID!

  \\"\\"\\"Fetches an object given its globally unique \`ID\`.\\"\\"\\"
  node(
    \\"\\"\\"The globally unique \`ID\`.\\"\\"\\"
    nodeId: ID!
  ): Node

  \\"\\"\\"Reads and enables pagination through a set of \`User\`.\\"\\"\\"
  allUsers(
    \\"\\"\\"Only read the first \`n\` values of the set.\\"\\"\\"
    first: Int

    \\"\\"\\"Only read the last \`n\` values of the set.\\"\\"\\"
    last: Int

    \\"\\"\\"
    Skip the first \`n\` values from our \`after\` cursor, an alternative to cursor
    based pagination. May not be used with \`last\`.
    \\"\\"\\"
    offset: Int

    \\"\\"\\"Read all values in the set before (above) this cursor.\\"\\"\\"
    before: Cursor

    \\"\\"\\"Read all values in the set after (below) this cursor.\\"\\"\\"
    after: Cursor

    \\"\\"\\"The method to use when ordering \`User\`.\\"\\"\\"
    orderBy: [UsersOrderBy!] = [PRIMARY_KEY_ASC]

    \\"\\"\\"
    A condition to be used in determining which values should be returned by the collection.
    \\"\\"\\"
    condition: UserCondition
  ): UsersConnection
  userById(id: Int!): User

  \\"\\"\\"Reads a single \`User\` using its globally unique \`ID\`.\\"\\"\\"
  user(
    \\"\\"\\"The globally unique \`ID\` to be used in selecting a single \`User\`.\\"\\"\\"
    nodeId: ID!
  ): User
}

input RegisterUserInput {
  name: String!
  email: String!
  bio: String
}

type RegisterUserPayload {
  user: User
}

type User implements Node {
  \\"\\"\\"
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  \\"\\"\\"
  nodeId: ID!
  id: Int!
  name: String!
  email: String!
  bio: String
  renamedComplexColumn: [Complex]
  createdAt: Datetime!
}

\\"\\"\\"
A condition to be used against \`User\` object types. All fields are tested for equality and combined with a logical ‘and.’
\\"\\"\\"
input UserCondition {
  \\"\\"\\"Checks for equality with the object’s \`id\` field.\\"\\"\\"
  id: Int

  \\"\\"\\"Checks for equality with the object’s \`name\` field.\\"\\"\\"
  name: String

  \\"\\"\\"Checks for equality with the object’s \`email\` field.\\"\\"\\"
  email: String

  \\"\\"\\"Checks for equality with the object’s \`bio\` field.\\"\\"\\"
  bio: String

  \\"\\"\\"Checks for equality with the object’s \`renamedComplexColumn\` field.\\"\\"\\"
  renamedComplexColumn: [ComplexInput]

  \\"\\"\\"Checks for equality with the object’s \`createdAt\` field.\\"\\"\\"
  createdAt: Datetime
}

\\"\\"\\"A connection to a list of \`User\` values.\\"\\"\\"
type UsersConnection {
  \\"\\"\\"A list of \`User\` objects.\\"\\"\\"
  nodes: [User]!

  \\"\\"\\"
  A list of edges which contains the \`User\` and cursor to aid in pagination.
  \\"\\"\\"
  edges: [UsersEdge!]!

  \\"\\"\\"Information to aid in pagination.\\"\\"\\"
  pageInfo: PageInfo!

  \\"\\"\\"The count of *all* \`User\` you could get from the connection.\\"\\"\\"
  totalCount: Int
}

\\"\\"\\"A \`User\` edge in the connection.\\"\\"\\"
type UsersEdge {
  \\"\\"\\"A cursor for use in pagination.\\"\\"\\"
  cursor: Cursor

  \\"\\"\\"The \`User\` at the end of the edge.\\"\\"\\"
  node: User
}

\\"\\"\\"Methods to use when ordering \`User\`.\\"\\"\\"
enum UsersOrderBy {
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  EMAIL_ASC
  EMAIL_DESC
  BIO_ASC
  BIO_DESC
  RENAMED_COMPLEX_COLUMN_ASC
  RENAMED_COMPLEX_COLUMN_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
}
"
`;

exports[`allows adding a simple mutation field to PG schema 2`] = `
Object {
  "user1": Object {
    "user": Object {
      "bio": null,
      "email": "testuser1@example.com",
      "id": "[id]",
      "name": "Test User 1",
      "nodeId": "[nodeId]",
    },
  },
  "user2": Object {
    "user": Object {
      "bio": "I have a bio!",
      "email": "testuser2@example.com",
      "id": "[id]",
      "name": "Test User 2",
      "nodeId": "[nodeId]",
    },
  },
}
`;
